sdk/go/common/workspace/plugins.go

```diff
@@ -344,6 +348,78 @@ func (source *gitlabSource) URL() string {
 	return fmt.Sprintf("gitlab://%s/%s", source.host, source.project)
 }
 
+type gitSource struct {
+	url  string
+	path string
+}
+
+func newGitSource(url *url.URL) (*gitSource, error) {
+	url.Scheme = "https"
+	u, path, err := gitutil.ParseGitRepoURL(url.String())
+	if err != nil {
+		return nil, err
+	}
+	return &gitSource{url: u, path: path}, nil
+}
+
+func (source *gitSource) GetLatestVersion(
+	ctx context.Context, _ func(*http.Request) (io.ReadCloser, int64, error),
+) (*semver.Version, error) {
+	// TODO: support this for the users convenience
+	return &semver.Version{},
+		errors.New("GetLatestVersion is not supported for git sources, please specify the version of the plugin")
+}
+
+func (source *gitSource) Download(
+	ctx context.Context, version semver.Version, opSy string, arch string,
+	_ func(*http.Request) (io.ReadCloser, int64, error),
+) (io.ReadCloser, int64, error) {
+	tmpdir := filepath.Join(os.TempDir(), "pulumi-plugins")
+	defer os.RemoveAll(tmpdir)
+	if version.Major == 0 && version.Minor == 0 && version.Patch == 0 && len(version.Pre) >= 0 {
+		if len(version.Pre) != 1 {
+			return nil, -1, fmt.Errorf("invalid version %s", version)
+		}
+		// The version string is prefixed with a 'x' character because Pre-versions can't
+		// start with a 0. Strip that off to get the actual hash.
+		hash := plumbing.NewHash(version.Pre[0].VersionStr[1:])
+		err := gitutil.GitCloneAndCheckoutCommit(source.url, hash, tmpdir)
+		if err != nil {
+			return nil, -1, err
+		}
+	} else {
+		var ref plumbing.ReferenceName
+		if version.Major == 0 && version.Minor == 0 && version.Patch == 0 {
+			ref = plumbing.HEAD
+		} else {
+			ref = plumbing.ReferenceName(version.String())
+		}
+		err := gitutil.GitCloneOrPull(source.url, ref, tmpdir, true /* shallow */)
+		if err != nil {
+			return nil, -1, err
+		}
+	}
+
+	tarbuf := bytes.NewBuffer([]byte{})
+	zip := gzip.NewWriter(tarbuf)
+
+	tw := tar.NewWriter(zip)
+
+	err := tw.AddFS(os.DirFS(filepath.Join(tmpdir, source.path)))
+	if err != nil {
+		return nil, -1, err
+	}
+
+	tw.Close()
+	zip.Close()
```

Comment:
Should these be `defer`ed before the potential error return?

Reply:
I wanted to do that, but we can't.  From https://go.dev/ref/spec#Defer_statements:

> deferred functions are executed after any result parameters are set by that return statement but before the function returns to its caller.

So this means that `tarbuf.Len()` would be calculated *before* we do `tw.Close()`, which in turn means that the length can be off.  

This is probably worth a comment though.